### ISSUE TITLE
C# Client may try to close a disposed WebSocket

### ISSUE THREAD

**Issue by moozzyk** [aspnet/SignalR#1004]
When the both - Send and Receive loops are complete the WebSockets transport will Dispose the client websocket it was using. Now, when someone calls `StopAsync` to stop the transport we will try to close the webSocket which results in an ObjectDisposedException. We do have a flag to skip closing the webSocket if it is in the Closed state but I found that the webSocket was in the CloseReceived state so the check did not prevent from trying to close the WebSocket. This can be reproduced end-to-end by calling DisposeAsync from the Closed event and shutting down the server: 

```C#
var hubConnection = new HubConnectionBuilder()
    .WithUrl("http://localhost:5000/chat")
    .WithTransport(TransportType.WebSockets)
    .WithConsoleLogger(Microsoft.Extensions.Logging.LogLevel.Trace)
    .Build();

var cts = new CancellationTokenSource();
hubConnection.Closed += async e =>
{
    await hubConnection.DisposeAsync();
    cts.Cancel();
};

await hubConnection.StartAsync();
await Task.Delay(-1, cts.Token);
```

Logs:
```
dbug: Microsoft.AspNetCore.Sockets.Client.HttpConnection[0]
      10/10/2017 11:44:41: Starting connection.
dbug: Microsoft.AspNetCore.Sockets.Client.HttpConnection[7]
      10/10/2017 11:44:41: Establishing Connection at: http://localhost:5000/chat.
dbug: Microsoft.AspNetCore.Sockets.Client.HttpConnection[2]
      10/10/2017 11:44:42: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Starting transport 'WebSocketsTransport' with Url: http://localhost:5000/chat?id=a7a8dd5c-3221-463b-bfb7-6f135b9eebaf.
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[0]
      10/10/2017 11:44:42: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Starting transport. Transfer mode: Text.
dbug: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[6]
      10/10/2017 11:44:42: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Starting the send loop.
dbug: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[2]
      10/10/2017 11:44:42: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Starting receive loop.
dbug: Microsoft.AspNetCore.Sockets.Client.HttpConnection[3]
      10/10/2017 11:44:42: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Raising Connected event.
trce: Microsoft.AspNetCore.Sockets.Client.HttpConnection[10]
      10/10/2017 11:44:42: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Beginning receive loop.
info: Microsoft.AspNetCore.SignalR.Client.HubConnection[22]
      Using HubProtocol 'json'.
dbug: Microsoft.AspNetCore.Sockets.Client.HttpConnection[17]
      10/10/2017 11:44:42: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Sending message.
dbug: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[12]
      10/10/2017 11:44:42: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Received message from application. Payload size: 20.
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[9]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Websocket closed by the server. Close status NormalClosure.
dbug: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[3]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Receive loop stopped.
trce: Microsoft.AspNetCore.Sockets.Client.HttpConnection[16]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Ending receive loop.
dbug: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[8]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Send loop canceled.
dbug: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[7]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Send loop stopped.
dbug: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[1]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Transport stopped.
dbug: Microsoft.AspNetCore.Sockets.Client.HttpConnection[4]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Ensuring all outstanding messages are processed.
dbug: Microsoft.AspNetCore.Sockets.Client.HttpConnection[5]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Draining event queue.
dbug: Microsoft.AspNetCore.Sockets.Client.HttpConnection[6]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Raising Closed event.
trce: Microsoft.AspNetCore.SignalR.Client.HubConnection[10]
      Shutting down connection.
info: Microsoft.AspNetCore.Sockets.Client.HttpConnection[18]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Stopping client.
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[5]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Transport is stopping.
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[15]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Closing WebSocket.
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[16]
      10/10/2017 11:44:49: Connection Id a7a8dd5c-3221-463b-bfb7-6f135b9eebaf: Closing webSocket failed.
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.WebSockets.ClientWebSocket'.
   at System.Net.WebSockets.ClientWebSocket.ThrowIfNotConnected()
   at Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport.<CloseWebSocket>d__21.MoveNext()
The client stopped with an error: A task was canceled.
```

Looks like there is even a simpler repro that does not require the Closed event handler - just call DisposeAsync after the websocket is closed on the server side.
```
info: Microsoft.AspNetCore.Sockets.Client.HttpConnection[18]
      10/10/2017 14:43:24: Connection Id bd98106e-ac28-4523-a84c-46e831a0f548: Stopping client.
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[5]
      10/10/2017 14:43:24: Connection Id bd98106e-ac28-4523-a84c-46e831a0f548: Transport is stopping.
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[15]
      10/10/2017 14:43:24: Connection Id bd98106e-ac28-4523-a84c-46e831a0f548: Closing WebSocket.
info: Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport[16]
      10/10/2017 14:43:24: Connection Id bd98106e-ac28-4523-a84c-46e831a0f548: Closing webSocket failed.
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.WebSockets.ClientWebSocket'.
   at System.Net.WebSockets.ClientWebSocket.ThrowIfNotConnected()
   at Microsoft.AspNetCore.Sockets.Client.WebSocketsTransport.<CloseWebSocket>d__21.MoveNext()
```

**Comment by BrennanConroy** [comment]
The comments in that section of the code indicate this is expected. We could reduce the logging level.
```
// Best effort - it's still possible (but not likely) that the transport is being closed via StopAsync
// while the webSocket is being closed due to an error.
```
and
```
// This is benign - the exception can happen due to the race described above because we would
// try closing the webSocket twice.
```

**Comment by moozzyk** [comment]
When I wrote this comment the situation I had in mind was a race where you happened to be closing the webSocket when it was being closed by the server (i.e. two independent calls that happened at roughly the same time) which is hard to avoid and hard to hit. At that time I did not think about scenario where you try to close the webSocket as a result of webSocket/transport/connection being closed. If you do that - as per my example - you will hit the exception 100% of time.
